### PR TITLE
Use jsonschema for rule config validation

### PR DESCRIPTION
**Pull request by nbrahms**
Previously we used an ersatz schema validation system scattered
throughout different parts of our code base, and requiring significant
manual work to update.

Here we substitute in jsonschema as a validation system. This should
allow us to catch more problems (e.g. the motivating example in #1460).

Advantages:
- Centralized schema definition using a publicly documented definition
  language
- Properly handles recursion in patterns, etc.
- Uniform errors
- Adds support for previously missing items (e.g. 'metavariable-regex')

Disadvantages:
- Adds an external dependency to our application

### CODE REVIEW

File: semgrep/semgrep/rule_lang.py [hunk 1]
```diff
@@ -328,11 +334,15 @@ def construct_object(self, node: Node, deep: bool = False) -> YamlTree:
     yaml = YAML()
     yaml.Constructor = SpanPreservingRuamelConstructor
     data = yaml.load(StringIO(contents))
+
+    schema_path = Path(__file__).parent / "rule_schema.yaml"
+    with schema_path.open() as fd:
+        validation_schema = yaml.load(fd).unroll()
```
Comment (mschwager) [inline semgrep/semgrep/rule_lang.py]: Can we store this as a const Python `dict` somewhere? That way we won't have to read something from disk every time we perform validation or include a file in the package MANIFEST.

File: semgrep/Pipfile [hunk 2]
```diff
@@ -12,6 +12,7 @@ appdirs = "*"
 pytest-xdist = "*"
 
 [packages]
+jsonschema = "~=3.2.0"
```
Comment (mschwager) [inline semgrep/Pipfile]: I believe we include dependencies in `setup.py` then rely on `path = "."` below for local installation.

File: semgrep/semgrep/rule_lang.py (hunk 1 above)
Comment (nbrahms) [reply semgrep/semgrep/rule_lang.py]: I much prefer using the external file, as we can publish it. I'm happy to load it once, though.

**Review by DrewDennison — APPROVED** [review]

**Comment by aryx** [comment]
Impressive! Far cleaner indeed.
Although this jsonschema specification is quite verbose ... specifying the same thing with yacc would be more compact (if it was possible).

**Pull request merged**